bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BigtableHBaseVeneerSettings.java

```diff
@@ -0,0 +1,570 @@
+      Preconditions.checkState(
+          !isNullOrEmpty(keyFileLocation), "service account location is null or empty");
+      credentials = GoogleCredentials.fromStream(new FileInputStream(keyFileLocation));
+
+    } else if (!isNullOrEmpty(configuration.get(BIGTABLE_SERVICE_ACCOUNT_EMAIL_KEY))) {
+
+      String serviceAccount = configuration.get(BIGTABLE_SERVICE_ACCOUNT_EMAIL_KEY);
+      LOG.debug("Service account %s specified.", serviceAccount);
+      String keyFileLocation = configuration.get(BIGTABLE_SERVICE_ACCOUNT_P12_KEYFILE_LOCATION_KEY);
+      Preconditions.checkState(
+          !isNullOrEmpty(keyFileLocation),
+          "Key file location must be specified when setting service account email");
+      LOG.debug("Using p12 keyfile: %s", keyFileLocation);
+
+      credentials = getCredentialFromPrivateKeyServiceAccount(serviceAccount, keyFileLocation);
+    }
+
+    if (credentials != null) {
+      stubSettings.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+    }
+  }
+
+  // copied over from CredentialFactory
+  // TODO: Find a better way to convert P12 key into Credentials instance
+  private Credentials getCredentialFromPrivateKeyServiceAccount(
+      String serviceAccountEmail, String privateKeyFile) throws IOException {
+    try {
+      PrivateKey privateKey =
+          SecurityUtils.loadPrivateKeyFromKeyStore(
+              SecurityUtils.getPkcs12KeyStore(),
+              new FileInputStream(privateKeyFile),
+              "notasecret",
+              "privatekey",
+              "notasecret");
+
+      return ServiceAccountJwtAccessCredentials.newBuilder()
+          .setClientEmail(serviceAccountEmail)
+          .setPrivateKey(privateKey)
+          .build();
+    } catch (GeneralSecurityException exception) {
+      throw new RuntimeException("exception while retrieving credentials", exception);
+    }
+  }
+
+  /** Creates {@link TransportChannelProvider} for plaintext negotiation type. */
+  private TransportChannelProvider buildPlainTextChannelProvider(String endpoint) {
+
+    InstantiatingGrpcChannelProvider.Builder channelBuilder =
+        defaultGrpcTransportProviderBuilder()
+            .setEndpoint(endpoint)
+            .setChannelConfigurator(
+                new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+                  @Override
+                  public ManagedChannelBuilder apply(ManagedChannelBuilder channelBuilder) {
+                    return channelBuilder.usePlaintext();
+                  }
+                });
+
+    String channelCount = configuration.get(BIGTABLE_DATA_CHANNEL_COUNT_KEY);
+    if (!isNullOrEmpty(channelCount)) {
+      channelBuilder.setPoolSize(Integer.parseInt(channelCount));
+    }
+
+    return channelBuilder.build();
+  }
+
+  /** Creates {@link Set} of {@link StatusCode.Code} from {@link Status.Code} */
+  private Set<StatusCode.Code> buildRetryCodes(Set<StatusCode.Code> retryableCodes) {
```

Comment:
I think this method should be split:
- extract extractRetryCodesFromConfig() that simply parses ADDITIONAL_RETRY_CODES from the config
- then use that function and ENABLE_GRPC_RETRIES_KEY  in buildIdempotentCallSettings()